crates/store/src/db/models/queries/accounts.rs

```diff
@@ -219,6 +220,119 @@ pub(crate) fn select_all_accounts(
+    pub last_block_included: BlockNumber,
+    /// Storage map values
+    pub values: Vec<StorageMapValue>,
+}
+
+impl StorageMapValue {
+    pub fn from_raw_row(row: (i64, i32, Vec<u8>, Vec<u8>)) -> Result<Self, DatabaseError> {
+        let (block_num, slot_index, key, value) = row;
+        Ok(Self {
+            block_num: BlockNumber::from_raw_sql(block_num)?,
+            slot_index: raw_sql_to_slot(slot_index),
+            key: Word::read_from_bytes(&key)?,
+            value: Word::read_from_bytes(&value)?,
+        })
+    }
+}
+
+/// Select account storage map values from the DB using the given [`SqliteConnection`].
+///
+/// # Returns
+///
+/// A vector of tuples containing `(slot, key, value, is_latest_update)` for the given account.
+/// Each row contains one of:
+///
+/// - the historical value for a slot and key specifically on block `block_to`
+/// - the latest updated value for the slot and key combination, alongside the block number in which
+///   it was updated
+pub(crate) fn select_account_storage_map_values(
+    conn: &mut SqliteConnection,
+    account_id: AccountId,
+    block_from: BlockNumber,
+    block_to: BlockNumber,
+) -> Result<StorageMapValuesPage, DatabaseError> {
+    use schema::account_storage_map_values as t;
+
+    // SELECT
+    //     block_num,
+    //     slot,
+    //     key,
+    //     value
+    // FROM
+    //     account_storage_map_values
+    // WHERE
+    //     account_id = ?1
+    //     AND block_num >= ?2
+    //     AND block_num <= ?3
+    // ORDER BY
+    //     block_num ASC
+    // LIMIT
+    //     :row_limit;
+
+    // TODO: These limits should be given by the protocol.
+    // See miden-base/issues/1770 for more details
+    pub const MAX_PAYLOAD_BYTES: usize = 2 * 1024 * 1024; // 2 MB
+    pub const ROW_OVERHEAD_BYTES: usize = size_of::<Word>() + size_of::<Word>() + size_of::<u8>(); // key + value + slot_idx
+    pub const ROW_LIMIT: usize = (MAX_PAYLOAD_BYTES / ROW_OVERHEAD_BYTES) + 1;
+
+    if !account_id.is_public() {
+        return Err(DatabaseError::AccountNotPublic(account_id));
+    }
+
+    if block_from > block_to {
+        return Err(DatabaseError::InvalidBlockRange { from: block_from, to: block_to });
+    }
+
+    let raw: Vec<(i64, i32, Vec<u8>, Vec<u8>)> =
+        SelectDsl::select(t::table, (t::block_num, t::slot, t::key, t::value))
+            .filter(
+                t::account_id
+                    .eq(account_id.to_bytes())
+                    .and(t::block_num.ge(block_from.to_raw_sql()))
+                    .and(t::block_num.le(block_to.to_raw_sql())),
+            )
+            .order(t::block_num.asc())
+            .limit(i64::try_from(ROW_LIMIT).expect("limit fits within i64"))
+            .load(conn)?;
+
+    // Discard the last block in the response (assumes more than one block may be present)
+    let (last_block_included, values) = if raw.len() == ROW_LIMIT {
```

Comment:
Nit: we sort of rely on sqlite's `LIMIT` implementation here


```suggestion
    let (last_block_included, values) = if raw.len() >= ROW_LIMIT {
```